circuit/src/merkle_tree/mod.rs

```diff
@@ -0,0 +1,297 @@
+/// Native version of the chip
+pub mod native;
+
+use std::marker::PhantomData;
+
+use crate::{
+	gadgets::set::{FixedSetChip, FixedSetConfig},
+	params::RoundParams,
+	poseidon::{PoseidonChip, PoseidonConfig},
+};
+use halo2::{
+	circuit::{AssignedCell, Layouter, Region},
+	halo2curves::FieldExt,
+	plonk::{Advice, Column, ConstraintSystem, Error},
+};
+
+const WIDTH: usize = 5;
+
+/// Configuration elements for the circuit are defined here.
+#[derive(Debug, Clone)]
+pub struct MerklePathConfig {
+	/// Configures an advice column for the temp.
+	temp: Column<Advice>,
+	/// Configures FixedSet circuit.
+	set: FixedSetConfig,
+	/// Configures Poseidon circuit.
+	poseidon: PoseidonConfig<WIDTH>,
+}
+
+/// Constructs a chip for the circuit.
+#[derive(Clone)]
+pub struct MerklePathChip<F: FieldExt, const LENGTH: usize, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// Constructs cell arrays for the nodes.
+	nodes: [[AssignedCell<F, F>; 2]; LENGTH],
+	/// Constructs a phantom data for the parameters.
+	_params: PhantomData<P>,
+}
+
+impl<F: FieldExt, const LENGTH: usize, P> MerklePathChip<F, LENGTH, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// Create a new chip.
+	pub fn new(nodes: [[AssignedCell<F, F>; 2]; LENGTH]) -> Self {
+		MerklePathChip { nodes, _params: PhantomData }
+	}
+
+	/// Make the circuit config.
+	pub fn configure(meta: &mut ConstraintSystem<F>) -> MerklePathConfig {
+		let temp = meta.advice_column();
+		let set = FixedSetChip::<F, 2>::configure(meta);
+		let poseidon = PoseidonChip::<_, WIDTH, P>::configure(meta);
+		meta.enable_equality(temp);
+		MerklePathConfig { temp, set, poseidon }
+	}
+
+	/// Synthesize the circuit.
+	pub fn synthesize(
+		&self, config: MerklePathConfig, path_arr: [[F; 2]; LENGTH], zero: AssignedCell<F, F>,
```

Comment:
Not sure why `path_arr` is needed. Its sufficient to check if hash of nodes at level `i` is one of the nodes at level `i + 1`. So we could use just `self.nodes` for this

Reply:
Ah, ok, `FixedSetChip` is not written properly. I will fix that at some point